Item Active Code + Config File update.

yasinallahdev opened the pull request:
Added the basic programming for item active code, including information that should allow them to respect swapping in the inventory, alongside being removed upon being sold. Life Steal is now loaded with items that have it as well, and the configuration files now respect Cooldown and Mana Cost settings.

#671

Commit:
Merge pull request #2 from LeagueSandbox/indev

Updating my fork to match head fork.

Commit:
Merge pull request #3 from LeagueSandbox/indev

Updating Repo.

Commit:
Item Active Code + Life Steal

Item Active data is now loaded when an item is purchased. Additionally, items with life steal not gained from a unique passive or aura (such as Doran's Blade or Bilgewater Cutlass) grant the stat to the item's purchaser.

Commit:
Merge pull request #4 from LeagueSandbox/indev

Actually updating for real this time.

Commit:
Config Files now respect Cooldown/Mana requirements.

Not much to add. Before, GameInfo.json's settings for Cooldowns and Mana Costs were not functional. These settings now serve there purpose; however, the static booleans in the Spell class have been removed due to redundancy.

MatthewFrench commented on GameServerLib/Content/ItemManager.cs:
```diff
@@ -164,6 +166,10 @@ public static ItemType Load(ItemManager owner, ItemContentCollectionEntry itemIn
                     FlatBonus = itemInfo.GetFloat("Data", "FlatCritDamageMod"),
                     PercentBonus = itemInfo.GetFloat("Data", "PercentCritDamageMod")
+                },
+                LifeSteel =
```
Heh typo

MatthewFrench reviewed (approved):
Looks and works great!

MythicManiac commented on GameServerLib/GameObjects/AttackableUnits/AI/Champion.cs:
```diff
@@ -536,6 +541,29 @@ public void UpdateSkin(int skinNo)
             Skin = skinNo;
         }
 
+        void IChampion.AddSpell(string name, byte slot, bool enabled)
```
Maybe call this `SetSpell` instead, as it overrides any existing spell from said slot.

MythicManiac commented on GameServerLib/GameObjects/Spells/Spell.cs:
```diff
@@ -67,6 +65,11 @@ public Spell(Game game, Champion owner, string spellName, byte slot)
             _spellGameScript.OnActivate(owner);
         }
 
+        void ISpell.Deactivate()
```
I personally prefer having access level defined (`private` / `public`) explicitly, but I'm not sure if we have a guideline on that 🤔

MythicManiac commented on GameServerLib/Items/Shop.cs:
```diff
@@ -47,6 +48,7 @@ public bool ItemBuyRequest(int itemId)
                 {
                     stats.RemoveModifier(item.ItemType);
                     _game.PacketNotifier.NotifyRemoveItem(_owner, inventory.GetItemSlot(item), 0);
+                    ((IChampion)_owner).RemoveSpell((byte)(inventory.GetItemSlot(item) + 6));
```
Instead of using the `6` as a magic number, we should have a constant somewhere that specifies where item spell slots start and use that here. Similar constant already exists for runes if I'm not mistaken.

FurkanS1821 replied:
or implement these functions for future use (pulled from the exe):
```cs
bool IsBasicAttackSlot(byte slot) => slot - 64 < 0x12;
bool IsBasicSpellSlot(byte slot) => slot < 4;
bool IsCastableItemSlot(byte slot) => slot - 4 < 7;
bool IsInternalSpellSlot(byte slot) => slot - 4 >= 0x28 && slot > 5 && slot - 64 > 0x11;
bool IsItemSlot(byte slot) => slot - 4 < 0x28;
bool IsPassiveSpellSlot(byte slot) => slot == 62;
bool IsUltimateSlot(byte slot) => slot == 3;
bool IsSlotless(byte slot) =>slot == 255;
```

MythicManiac replied:
@FurkanS1821 you should create an issue for that, and also mention the use of constants 😉

FurkanS1821 replied:
yeah ur right

MythicManiac reviewed (changes requested):
I would like to see a few changes as explained in the comments before merging. Good work so far regardless 👍

Commit:
Updated Code as per Request.

Updated code for certain naming requests, access requirements, and use of constants.

MythicManiac reviewed (approved):
Looks good 👍 nice work

Commit:
Merge branch 'indev' into indev

Commit:
Removed the 'CanChampionMove' change.

MythicManiac reviewed (approved):
LGTM 👍

MythicManiac commented on GameServerLib/GameObjects/AttackableUnits/AI/Champion.cs:
```diff
@@ -222,14 +222,6 @@ public Vector2 GetRespawnPosition()
             return new Vector2(coords.X, coords.Y);
         }
 
-        public void StopChampionMovement()
```
Actually this shouldn't be removed 🤔 github messed up

MythicManiac reviewed (changes requested):
On second pass, github derped with the diff and the last commit undid the derp, and now there's a removal in the diff `StopChampionMovement` that shouldn't be there. Once that's removed (the removal), this is good to go 👍 Thanks for all the effort so far 😄

Commit:
Revert "Removed the 'CanChampionMove' change."

This reverts commit 8e41ea5df360cbc1be4dfc09eb692c6332e0147a.

MythicManiac reviewed (approved):
Looks good 👍 nice work
